### PR TITLE
LargeCount MPI_write/read_at

### PR DESCRIPTION
**Pull request by zjiaqi2018**
Use `dealii::Utilities::MPI::LargeCount::MPI_File_write/read_at_c`.
Similar to https://github.com/dealii/dealii/pull/13851

### CODE REVIEW

File: source/distributed/fully_distributed_tria.cc [hunk 1]
```diff
@@ -561,13 +562,13 @@ namespace parallel
...
-        ierr = MPI_File_write_at(fh,
-                                 mpisize * sizeof(unsigned int) +
-                                   offset, // global position in file
-                                 buffer.data(),
-                                 buffer.size(), // local buffer
-                                 MPI_CHAR,
-                                 MPI_STATUS_IGNORE);
+        ierr = dealii::Utilities::MPI::LargeCount::MPI_File_write_at_c(
+          fh,
+          mpisize * sizeof(unsigned int) + offset, // global position in file
```
Comment (tjhei) [inline source/distributed/fully_distributed_tria.cc:567]: Can you compute this position before and store it in a 64bit int?

**Review by tjhei — COMMENTED** [review]
Thank you, but I think the ``offset`` stored at the beginning of the file also needs to be 64 bits long so this whole code needs fixing to make it 64bit safe.

**Comment by tjhei** [comment]
Let me know if you need more guidance.

File: source/distributed/fully_distributed_tria.cc (hunk 1 above)
Comment (zjiaqi2018) [reply source/distributed/fully_distributed_tria.cc:567]: like this?

```
MPI_Offset offset = 0;

MPI_Offset global_position = mpisize * sizeof(unsigned int) + offset;
```

File: source/distributed/fully_distributed_tria.cc [hunk 2]
```diff
@@ -539,35 +540,37 @@ namespace parallel
         dealii::Utilities::pack(construction_data, buffer, false);
 
         // Write offsets to file.
-        unsigned int buffer_size = buffer.size();
+        std::uint64_t buffer_size = buffer.size();
```
Comment (tjhei) [inline source/distributed/fully_distributed_tria.cc:543]: const

File: source/distributed/fully_distributed_tria.cc [hunk 3]
```diff
@@ -539,35 +540,37 @@ namespace parallel
...
-                                 myrank * sizeof(unsigned int),
+                                 myrank * sizeof(std::uint64_t),
                                  &buffer_size,
                                  1,
-                                 MPI_UNSIGNED,
+                                 MPI_UINT64_T,
                                  MPI_STATUS_IGNORE);
         AssertThrowMPI(ierr);
 
+        std::uint64_t global_position =
```
Comment (tjhei) [inline source/distributed/fully_distributed_tria.cc:564]: const

File: source/distributed/fully_distributed_tria.cc [hunk 4]
```diff
@@ -539,35 +540,37 @@ namespace parallel
...
-        ierr = MPI_File_write_at(fh,
-                                 mpisize * sizeof(unsigned int) +
-                                   offset, // global position in file
-                                 buffer.data(),
-                                 buffer.size(), // local buffer
-                                 MPI_CHAR,
-                                 MPI_STATUS_IGNORE);
+        ierr = dealii::Utilities::MPI::LargeCount::MPI_File_write_at_c(
+          fh,
+          global_position, // global position in file
```
Comment (tjhei) [inline source/distributed/fully_distributed_tria.cc:569]: The comment here can now go. That is why I suggested the new variable, by the way. Comment through code (variable names) is much cleaner than what was done here.

File: source/distributed/fully_distributed_tria.cc [hunk 5]
```diff
@@ -673,13 +676,13 @@ namespace parallel
 
         // Read buffers from file.
         std::vector<char> buffer(buffer_size);
-        ierr = MPI_File_read_at(fh,
```
Comment (tjhei) [inline source/distributed/fully_distributed_tria.cc:676]: You will also adjust all the read stuff to 64 bit (above).

File: source/distributed/fully_distributed_tria.cc [hunk 6]
```diff
@@ -673,13 +676,13 @@ namespace parallel
...
-                                mpisize * sizeof(unsigned int) +
-                                  offset, // global position in file
-                                buffer.data(),
-                                buffer.size(), // local buffer
-                                MPI_CHAR,
-                                MPI_STATUS_IGNORE);
+        ierr = dealii::Utilities::MPI::LargeCount::MPI_File_read_at_c(
+          fh,
+          mpisize * sizeof(unsigned int) + offset, // global position in file
```
Comment (tjhei) [inline source/distributed/fully_distributed_tria.cc:681]: Introduce a variable here as well please.

**Review by tjhei — COMMENTED** [review]
The reading part will need to be fixed as well. Do you know if we have a test for this piece of code?

**Comment by zjiaqi2018** [comment]
> The reading part will need to be fixed as well. Do you know if we have a test for this piece of code?

I think we have one test: https://github.com/dealii/dealii/blob/master/tests/fullydistributed_grids/save_load_01.cc

**Comment by tjhei** [comment]
Does is pass? :-)

**Comment by zjiaqi2018** [comment]
> Does is pass? :-)

Yes!

**Comment by drwells** [comment]
This conflicts with #13885 - we'll have to rebase one or the other based on the merge order.

File: source/distributed/fully_distributed_tria.cc [hunk 7]
```diff
@@ -539,35 +540,39 @@ namespace parallel
...
         AssertThrowMPI(ierr);
 
         // Write offsets to file.
         ierr = MPI_File_write_at(fh,
-                                 myrank * sizeof(unsigned int),
+                                 myrank * sizeof(std::uint64_t),
                                  &buffer_size,
                                  1,
-                                 MPI_UNSIGNED,
+                                 MPI_UINT64_T,
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc:560]: A slightly better alternative would have been to use this instead:
```suggestion
                                 Utilities::MPI::mpi_type_id_for_type<decltype(buffer_size)>,
```
to make sure that type and variable remain in sync.

File: source/distributed/fully_distributed_tria.cc [hunk 8]
```diff
@@ -651,35 +656,39 @@ namespace parallel
...
-        unsigned int buffer_size;
+        std::uint64_t buffer_size;
 
         ierr = MPI_File_read_at(fh,
-                                myrank * sizeof(unsigned int),
+                                myrank * sizeof(std::uint64_t),
                                 &buffer_size,
                                 1,
-                                MPI_UNSIGNED,
+                                MPI_UINT64_T,
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc:665]: Same here

File: source/distributed/fully_distributed_tria.cc [hunk 9]
```diff
@@ -651,35 +656,39 @@ namespace parallel
...
         AssertThrowMPI(ierr);
 
-        unsigned int offset = 0;
+        std::uint64_t offset = 0;
 
         ierr = MPI_Exscan(&buffer_size,
                           &offset,
                           1,
-                          MPI_UNSIGNED,
+                          MPI_UINT64_T,
```
Comment (bangerth) [inline source/distributed/fully_distributed_tria.cc:674]: And here

**Review by bangerth — APPROVED** [review]
Looks good to me. We can merge this as-is and improve the type deduction separately.

**Comment by bangerth** [comment]
@tjhei Your call on whether this is ready to merge!

**Comment by tjhei** [comment]
This needs a rebase after we merge #13885.

**Review by tjhei — COMMENTED** [review]
This looks good to me. This is a strict improvement over the current status and I am fairly confident this is correct. Let's merge #13885, rebase, and adjust the function names, though.

**Review by kronbichler — APPROVED** [review]

File: source/distributed/fully_distributed_tria.cc [hunk 10]
```diff
@@ -539,35 +540,39 @@ namespace parallel
...
+
         // Write buffers to file.
-        ierr = MPI_File_write_at(fh,
-                                 mpisize * sizeof(unsigned int) +
-                                   offset, // global position in file
-                                 buffer.data(),
-                                 buffer.size(), // local buffer
-                                 MPI_CHAR,
-                                 MPI_STATUS_IGNORE);
+        ierr = dealii::Utilities::MPI::LargeCount::MPI_File_write_at_c(
```
Comment (drwells) [inline source/distributed/fully_distributed_tria.cc:569]:
```suggestion
        ierr = dealii::Utilities::MPI::LargeCount::File_write_at_c(
```

File: source/distributed/fully_distributed_tria.cc [hunk 11]
```diff
@@ -651,35 +656,39 @@ namespace parallel
...
         // Read buffers from file.
         std::vector<char> buffer(buffer_size);
-        ierr = MPI_File_read_at(fh,
-                                mpisize * sizeof(unsigned int) +
-                                  offset, // global position in file
-                                buffer.data(),
-                                buffer.size(), // local buffer
-                                MPI_CHAR,
-                                MPI_STATUS_IGNORE);
+        ierr = dealii::Utilities::MPI::LargeCount::MPI_File_read_at_c(
```
Comment (drwells) [inline source/distributed/fully_distributed_tria.cc:685]:
```suggestion
        ierr = dealii::Utilities::MPI::LargeCount::File_read_at_c(
```

**Review by drwells — CHANGES_REQUESTED** [review]
I recommend rebasing to make sure we are up-to-date with the change in the other large count PR.

**Review by tjhei — APPROVED** [review]
Looks good. Can you squash your commits, please?

**Review by kronbichler — APPROVED** [review]

**Pull request merged**